### ISSUE TITLE
Question: array of arrays

### ISSUE THREAD

**Issue by gergoerdi** [Cryptjar/avr-progmem-rs#8]
I'd like to store sprite animation frames in `PROGMEM`, where each frame has the same size (represented as a `[u8; 8]`) but the number of frames can vary between animations.

So I have something like

```
progmem! {
  pub static progmem IDLE: [[u8; 8]; 1] = ....;
  pub static progmem WALK: [[u8; 8]; 3] = ...;
```

The first issue is dereferencing these. I can use e.g. `WALK.load_at(2)` to get a `[u8; 8]`, but if I want to process it byte-wise, is there a way to avoid that 7-byte overhead and get just the one byte, something like e.g.  `WALK.load_wrapper_at(2).load_at(3)`?

The second issue is storing a reference to the currently playing animation. If they were all the same length, I could store a `&'static ProgMem<[[u8;8];3>` in my state; however, since they are not, one of the wrappers has type `ProgMem<[[u8;8];1>` while the other one is `ProgMem<[[u8;8];3>`. I think what I'd need is something like a `ProgMemArrayRef<[u8;8]>` that has a dynamic length accessible at runtime.

**Comment by gergoerdi** [comment]
For the first issue, I guess I could instead store these as `[u8; 8 * 1]` and `[u8; 8 * 3]` respectively, and do some index arithmetic before `load_at`; but this doesn't solve the second issue.

**Comment by Cryptjar** [comment]
Hi, thanks for sharing your interesting use-case.

Regarding your first point, I think, it should be quite possible to add an method that returns a `ProgMem<T>` from a `ProgMem<[T;N]>` at an arbitrary index, without loading the entire thing yet. Thus it should then be possible to access single elements for multi-dimensional arrays.

For your second point, I'm not sure how to support it easily. The option I can think of so far are, either:
* lift the sized-constraint on `PromMem` and add conversion and access methods for slices, or
* add something like a `ProgMemSlice`, mirroring mostly the `PromMem` API

I'm not sure if the first option will work with today's Rust, and the second option seems we would end up with the entire API twice.

I'll play a bit around and see what works.

**Comment by Cryptjar** [comment]
Looks like lifting the size-constraint is the way to go. I started implementing it on the [`impl-unsized` branch](https://github.com/Cryptjar/avr-progmem-rs/tree/impl-unsized) if you want to try it out.

Basically, you can now coerce a wrapper of an array of any size into a wrapper of a slice:

```rust
progmem!{
    static progmem ARRAY: [u8; 3] = [1,2,3];
}
// The array wrapper
let array: ProgMem<[u8; 3]> = ARRAY;
// Coerced to a slice wrapper, just like that.
let slice: ProgMem<[u8]> = array;
```

Thus, you get the same type independent of the size of the original array.

Then on these slices, you can use the `at` and `loat_at` methods as usual.

**Comment by gergoerdi** [comment]
> I started implementing it on the [`impl-unsized` branch](https://github.com/Cryptjar/avr-progmem-rs/tree/impl-unsized) if you want to try it out.

I think it's working very nicely!

**Comment by Cryptjar** [comment]
Released as part of `v0.3.2`

**Comment by gergoerdi** [comment]
Thanks!
